### PR TITLE
binning API now can return a factor column or an integer one

### PR DESCRIPTION
**Pull request by nhanitvn**
### Description and related tickets, documents
- That is done by adding **toLabels** params to allow the output column to be an integer one which contains zero-based indices instead of a factor column of range strings
- **Refactor the code, move the implementation to ABinningHandler to avoid duplicated code in spark/jdbc BinningHandler**
- Add convenient methods for binning to DDF while keeping the old one for backward compatible
- Add precision param for number of decimal digits to display in range strings (default is 2)
- Automatically search for appropriate precision value (20 tries from the specified precision value, default is 2) for nice labels
- Add validation for duplicated/non-monotonically-increasing breaks
- Binning by equal intervals and by equal frequency always use includeLowest = True despite any supplied value
- In case of binning by equal intervals, if min = max, the lower and upper bounds are extended by .1% as in Pandas and R to ensure that we can still map values to a range/value.
- Change EQUAlFREQ (lower-cased L) to EQUALFREQ (upper-cased L)

Reviewers: @hai-adatao @phvu @Huandao0812 

### How to test
Describe how this PR is tested. In case manual testing is required, describe how to do so.

### PR Progress
Make sure all checkboxes below are checked before merged
- [x] Branch is in format `prefix/description` (see [this](http://www.guyroutledge.co.uk/blog/git-branch-naming-conventions/))
- [x] Merge check has no conflicts. PR checks passed.
- [ ] Code review is done.

### CODE REVIEW

File: core/src/main/java/io/ddf/analytics/ABinningHandler.java [hunk 1]
```diff
@@ -22,21 +30,192 @@ public ABinningHandler(DDF theDDF) {
...
+    BinningType binType = BinningType.get(binningType);
+
+    double[] theBreaks = breaks;
+
+    switch(binType) {
+      case CUSTOM:
+        if (breaks == null) throw new DDFException("Please enter valid break points");
+        // check for monotonic
+        Arrays.sort(theBreaks);
+        if (!Arrays.equals(theBreaks, breaks)) throw new DDFException("Please enter increasing breaks");
```
Comment (phvu) [inline core/src/main/java/io/ddf/analytics/ABinningHandler.java]: Does this work? `Arrays.sort()` is in-place, and `theBreaks` is just a reference to `breaks`, so after this `Arrays.equals(theBreaks, breaks)` will be true anyway?
You might need to clone `breaks` somehow.

Comment (nhanitvn) [inline core/src/main/java/io/ddf/analytics/ABinningHandler.java]: Good point, let me write a test for that.

File: core/src/main/java/io/ddf/analytics/ABinningHandler.java [hunk 2]
```diff
@@ -22,21 +30,192 @@ public ABinningHandler(DDF theDDF) {
...
+        if (numBins < 2) throw new DDFException("Number of bins cannot be smaller than 2");
+        theBreaks = getQuantilesFromNumBins(colMeta.getName(), numBins);
+        includeLowest = true;
+        break;
+      case EQUALINTERVAL:
+        if (numBins < 2) throw new DDFException("Number of bins cannot be smaller than 2");
+        theBreaks = getIntervalsFromNumBins(colMeta.getName(), numBins, right);
+        includeLowest = true;
+        break;
+      default:throw new DDFException(String.format("Binning type %s is not supported", binningType));
```
Comment (phvu) [inline core/src/main/java/io/ddf/analytics/ABinningHandler.java]: nit:  new line after `default:`

File: core/src/main/java/io/ddf/analytics/ABinningHandler.java [hunk 3]
```diff
@@ -22,21 +30,192 @@ public ABinningHandler(DDF theDDF) {
...
+    }
+
+    // Check for uniqueness
+    List<Double> listBreaks = Arrays.asList(ArrayUtils.toObject(theBreaks));
+    if (new HashSet<>(listBreaks).size() < theBreaks.length) {
```
Comment (phvu) [inline core/src/main/java/io/ddf/analytics/ABinningHandler.java]: One java8 way to do this is: `if (!Arrays.stream(theBreaks).allMatch(new HashSet<>()::add)) {`

File: core/src/main/java/io/ddf/analytics/ABinningHandler.java [hunk 4]
```diff
@@ -22,21 +30,192 @@ public ABinningHandler(DDF theDDF) {
...
 
-    newddf.getMetaDataHandler().copyFactor(this.getDDF());
-    return newddf;
+    newDDF.getMetaDataHandler().copyFactor(this.getDDF());
+    return newDDF;
+  }
+
+  private double[] getIntervalsFromNumBins(String colName, int numBins, Boolean right) throws DDFException {
+    DDF ddf = this.getDDF();
+    double[] minMax = new double[] {ddf.getVectorMin(colName), ddf.getVectorMax(colName)};
```
Comment (phvu) [inline core/src/main/java/io/ddf/analytics/ABinningHandler.java]: nit: unnecessary array

**Comment by hai-adatao** [comment]
@nhanitvn can you help put all these tests in `ddf-test/BinningHandlerBaseSuite`, and then implement spark test suite instead? Check out `JoinHandlerBaseSuite` and `JoinHandlerSuite` for an example.
Also, would appreciate if you take a look at the tests in `ddf-test/BinningHandlerBaseSuite`, they're all ignored, if they're not valid anymore please help remove it, or refactor it otherwise.

**Comment by hai-adatao** [comment]
Let's try to get this in 1.4.17

File: core/src/main/java/io/ddf/analytics/ABinningHandler.java (hunk 3 above)
Comment (nhanitvn) [inline core/src/main/java/io/ddf/analytics/ABinningHandler.java]: Wow, concise code. However it seems to be magical and difficult to figure out how it achieve the goal. May be we just keep the current one.

**Pull request merged**